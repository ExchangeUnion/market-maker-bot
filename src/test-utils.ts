import { Config } from './config';
import { Level, Loggers } from './logger';

const testConfig = (): Config => {
  return {
    LOG_LEVEL: Level.Trace,
    CEX: 'Binance',
    CEX_API_KEY: '123',
    CEX_API_SECRET: 'abc',
    DATA_DIR: '',
    OPENDEX_CERT_PATH: `${__dirname}/../mock-data/tls.cert`,
    OPENDEX_RPC_HOST: 'localhost',
    OPENDEX_RPC_PORT: '1234',
    MARGIN: '0.06',
    BASEASSET: 'ETH',
    QUOTEASSET: 'BTC',
    CEX_BASEASSET: 'ETH',
    CEX_QUOTEASSET: 'BTC',
    TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: '321',
    TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: '123',
    TEST_MODE: true,
  };
};

const getLoggers = (): Loggers => {
  const mockLogger = {
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
    trace: () => {},
    error: () => {},
  };
  return ({
    global: mockLogger,
    centralized: mockLogger,
    opendex: mockLogger,
  } as unknown) as Loggers;
};

type TestError = {
  code: string | number;
  message: string;
};

export { getLoggers, testConfig, TestError };
