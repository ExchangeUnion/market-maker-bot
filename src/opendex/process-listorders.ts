import { ListOrdersResponse, Order } from '../proto/xudrpc_pb';
import { Config } from 'src/config';
import { errors } from './errors';

type ProcessListOrdersParams = {
  config: Config;
  listOrdersResponse: ListOrdersResponse;
};

const processListorders = ({
  listOrdersResponse,
  config,
}: ProcessListOrdersParams): string[] => {
  const ordersMap = listOrdersResponse.getOrdersMap();
  const { BASEASSET, QUOTEASSET } = config;
  const tradingPair = `${BASEASSET}/${QUOTEASSET}`;
  const tradingPairOrders = ordersMap.get(tradingPair);
  if (!tradingPairOrders) {
    throw errors.INVALID_ORDERS_LIST(tradingPair);
  }
  const buyOrders = tradingPairOrders.getBuyOrdersList();
  const sellOrders = tradingPairOrders.getSellOrdersList();
  const orderIds = buyOrders
    .concat(sellOrders)
    .reduce((allOrderIds: string[], currentOrder: Order) => {
      const ownOrder = currentOrder.getIsOwnOrder();
      const quantity = currentOrder.getQuantity();
      const hold = currentOrder.getHold();
      const remainingQuantity = quantity - hold;
      if (ownOrder && remainingQuantity) {
        return allOrderIds.concat(currentOrder.getLocalId());
      }
      return allOrderIds;
    }, []);
  return orderIds;
};

export { processListorders, ProcessListOrdersParams };
