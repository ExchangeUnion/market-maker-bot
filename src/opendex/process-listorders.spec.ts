import { equals } from 'ramda';
import { v4 as uuidv4 } from 'uuid';
import { ListOrdersResponse } from '../proto/xudrpc_pb';
import { testConfig } from '../test-utils';
import { processListorders } from './process-listorders';

type TestOrderParams = {
  quantity: number;
  hold: number;
  ownOrder?: boolean;
};

type TestOrder = {
  getLocalId: () => string;
  getQuantity: () => number;
  getHold: () => number;
  getIsOwnOrder: () => boolean;
};

const testOrder = ({
  quantity,
  hold,
  ownOrder = true,
}: TestOrderParams): TestOrder => {
  const orderId = uuidv4();
  return {
    getLocalId: () => orderId,
    getQuantity: () => quantity,
    getHold: () => hold,
    getIsOwnOrder: () => ownOrder,
  };
};

type AssertProcessListOrdersParams = {
  buyOrders: TestOrder[];
  sellOrders: TestOrder[];
  expectedIds: string[];
};

const assertProcessListOrders = ({
  buyOrders,
  sellOrders,
  expectedIds,
}: AssertProcessListOrdersParams) => {
  const ordersMap = new Map();
  const orders = {
    getBuyOrdersList: () => buyOrders,
    getSellOrdersList: () => sellOrders,
  };
  const { BASEASSET, QUOTEASSET } = testConfig();
  ordersMap.set(`${BASEASSET}/${QUOTEASSET}`, orders);
  const listOrdersResponse = ({
    getOrdersMap: () => ordersMap,
  } as unknown) as ListOrdersResponse;
  const orderIds = processListorders({
    config: testConfig(),
    listOrdersResponse,
  });
  const orderIdsMatch = equals(orderIds, expectedIds);
  expect(orderIdsMatch).toBeTruthy();
};

describe('processListorders', () => {
  it('returns list of order ids from ListOrdersResponse', () => {
    expect.assertions(1);
    const sellOrder = testOrder({
      quantity: 100000,
      hold: 0,
    });
    const buyOrder = testOrder({
      quantity: 123000,
      hold: 0,
    });
    const expectedIds = [buyOrder.getLocalId(), sellOrder.getLocalId()];
    assertProcessListOrders({
      sellOrders: [sellOrder],
      buyOrders: [buyOrder],
      expectedIds,
    });
  });

  it('ignores orders with all of quantity on hold', () => {
    expect.assertions(1);
    const sellOrder = testOrder({
      quantity: 100000,
      hold: 100000,
    });
    const buyOrder = testOrder({
      quantity: 123000,
      hold: 123000,
    });
    assertProcessListOrders({
      sellOrders: [sellOrder],
      buyOrders: [buyOrder],
      expectedIds: [],
    });
  });

  it("ignores others' orders", () => {
    expect.assertions(1);
    const sellOrder = testOrder({
      quantity: 100000,
      hold: 0,
      ownOrder: false,
    });
    const buyOrder = testOrder({
      quantity: 123000,
      hold: 0,
      ownOrder: false,
    });
    assertProcessListOrders({
      sellOrders: [sellOrder],
      buyOrders: [buyOrder],
      expectedIds: [],
    });
  });

  it('errors when orders list is missing', () => {
    expect.assertions(1);
    const ordersMap = new Map();
    const listOrdersResponse = ({
      getOrdersMap: () => ordersMap,
    } as unknown) as ListOrdersResponse;
    expect(() => {
      processListorders({
        config: testConfig(),
        listOrdersResponse,
      });
    }).toThrowErrorMatchingSnapshot();
  });
});
