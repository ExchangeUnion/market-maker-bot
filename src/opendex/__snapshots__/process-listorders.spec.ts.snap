// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processListorders errors when orders list is missing 1`] = `"Could not retrieve orders list for ETH/BTC"`;
